### DISCUSSION TITLE
Are optimistic updates meant to work for adding items to lists?

### DISCUSSION THREAD

**Discussion by erquhart** [urql-graphql/urql#2470] [Q&A] [answered]
I have optimistic updates working for updates and deletions, but they simply won't work when creating a new item in a list.

They also result in the update never occurring at all - have to refresh the screen.

I've ensured the optimistic update returns exactly the same information that eventually comes back from the API, as well as matching all fields on the query.

Wondering if this is by design? The only use case mentioned is updating existing fields (e.g, likes/stars). Anyone succesfully using optimistics for create operations?

**Comment by JoviDeCroock** [comment, marked as answer]
Should just work if you have an updater that works for that mutation-field that does something like 

```
updates: {
  Mutation: {
    addX: () => {
            cache.updateQuery(
          { query: YOUR_LIST },
          (data) => {
            // push stuff

            return data;
    }
  }
}
```

**Reply by erquhart** [reply]
Okay, that's good to know.

I do have updaters for my create mutations, but when I add an optimistic something different happens and the updater is unsuccessful. The request succeeds, but the response never gets into the cache, even though my updater is adding it (both for the optimistic and the actual response). I've logged it out - the updater is adding an identical object to the cache with and without the optimistic, there are zero differences, so I'm guessing urql is handling the optimistic differently somehow.

In case anyone can spot something I'm missing, here's an example:

```js
optimistic: {
  createEvent: (args, cache, info) => {
    const { input } = info.variables
    return {
      __typename: 'Event',
      amount: null,
      autopay: null,
      date: null,
      note: null,
      status: null,
      dateExpected: null,
      ...input,
    }
  }
},
updates: {
  Mutation: {
    createEvent(result, _args, cache) {
      const scheduleId = result.createEvent.scheduleId
      cache.updateQuery(
        {
          query: gql`
            query ($scheduleId: uuid!) {
              schedule(id: $scheduleId) {
                id
                events {
                  id
                }
              }
            }
          `,
          variables: { scheduleId },
        },
        (data) => {
          if (!data) {
            return {
              schedule: {
                __typename: 'Schedule',
                id: scheduleId,
                events: [result.createEvent],
              },
            }
          }
          data.schedule.events.push(result.createEvent)
          return data
        }
      )
    }
  }
}
```

**Reply by JoviDeCroock** [reply]
is result.createEvent.__typename conveying the correct information? Else a small repro might help us help you

**Reply by erquhart** [reply]
Yeah, typename is coming out accurate. Walking away from this after a day of tears, I’ll get a repro case up when I circle back. Thanks so much for your responses.

**Reply by nandorojo** [reply]
@erquhart In my experience, even though the docs say otherwise, you do indeed need to update the data immutably. Or at least it's worth trying.

**Reply by erquhart** [reply]
@JoviDeCroock very late follow up as this still pops back up every now and then: you mentioned using `updateQuery` here, which does work, but is there any reason you'd expect `cache.link` to not work just as well?

**Reply by kitten** [reply]
@erquhart: `cache.link` will work just as well here, and should in any case. Do let me know if you run into anything unexpected.

I'm currently thinking about how to make mistakes in optimistic updates/resolvers more obvious. For instance, by somehow issuing warnings for cache misses after updates.

`cache.link` will always work just as well to update a list of relations, and is preferred, provided that all entities you're pointing to are actually already in the cache. It will _not_ work, if your mutation does not actually write the item to the cache.

The failure cases are as follows:
- `cache.link` writes a new relation (or relations) but one or more entities aren't cached at all
- if the above isn't true; the mutation writes your entity and `cache.link` establishes the relation, but a query's selection (i.e. data dependencies) are _greater_ than what has been written by the mutation
- if the above isn't true; this should work.

These are (in theory) the only two ways in which `cache.link` can fail in a mutation update, be it optimistic or not. If the mutation _isn't_ optimistic, you'll see a new query being sent. If the mutation _is_ optimistic, we can't issue a query, since it'd conflict (a) with the mutation, and (b) with your optimistic update.

This is the less obvious failure case. In this second case, the query will wait until the mutation has resolved and may then refetch, if there's still a cache miss.

However, sometimes it won't refetch, because the first failure case means that the relation was never established.

(in a way this is "good" news, because both failure cases are consistent, and the first will never occur once the problem is fixed in the cache configuration)

**Reply by erquhart** [reply]
So given we have foos, bars, and fooBars, with this query and mutation:
```
query {
  foo(id: $id) {
    id
    fooBars {
      id
      bar {
        id
        name
      }
    }
  }
}

mutation {
  createFooBar {
    id
    fooId
    barId
  }
}
```

If I do `createFooBar` and the `bar` is already in the cache, but my mutation isn't writing or updating the `bar` entity, the link will fail?

I'm seeing:
- `updateQuery` is working for directly updating this specific query, but that's arduous and tough at scale
- with no updater, everything works automatically
- given this last point, I'd expect simply adding an optimistic update to work, but it seems the updater is required

**Reply by kitten** [reply]
I'm not quite sure I understand what you mean, sorry!

Given the above example, you'll just run `createFooBar`, which updates nothing if it just contains IDs, except for the entity itself.

I'm also not sure then what you mean by no updater, or what you're expecting, or whether you maybe have resolvers in place 😅 

If you add nothing and just have this mutation with Graphcache and no config in place, then after the mutation nothing will happen, and nothing will update, except for the `FooBar` if it's already present somewhere, which it presumably can't be since the mutation is creating it.

**Reply by erquhart** [reply]
Yeah I guess that's a lot of gaps. I'll try again coming off of your reasons cache.link would fail:

You're saying it fails for a query who's selection is greater than _what the mutation writes_, but I'm guessing that previously cached data also works right?.  So if the mutation is just creating a many-to-many relationship, and both sides of the relationship are already cached, I should only need to link the two entities to the new entity and the update should work. Does that sound right?

**Reply by kitten** [reply]
It's basixcally like that, but I'm not sure whether maybe your mental model of this might be more complicated than necessary? 🤔 As per the docs, this is the relevant diagram:

<img src="https://formidable.com/open-source/urql/static/query-document-info.8eae17e1.png" alt="Diagram: Normalization is based on types, keys, and relations. This information can all be inferred from the query document." width="350"/>

So, every entity (i.e. an instance of an object type in your schema, in other words an object with a `__typename` field) has a key. That's even if it has no keyable field (embedded data). Then every field has a key.

Graphcache has two stores that look like lookup tables. In one table, we store scalars aka records, like numbers and stuff, i.e. stuff without `__typename`. In the other table, we store relations aka links. Relations are just other entity keys. When we look up an entry, it's looked up by the object's key and the field key. A field key is just the field name + arguments (if any).

So, if we apply this to all GraphQL query documents you could write. Every single field in that document, will have a field key. Every time we have a sub-selection set, we have a relation/link.

Assuming each entity has a key, and we ignore embedded data, the problem is that mutations point to arbitrary entities and may leave out data. For instance, your mutation of course contains no list.

Cache updates must be written for each relation that a mutation can't "see" or rather doesn't contain. So for instance, a list field can't know that it contains a new item a mutation creates, unless it has a special resolver or updater.

tl;dr, it doesn't matter which kind of relation a mutation creates. You should update relations (or invalidate them) that the mutation doesn't contain.

I hope that's more comprehensive for future readers too 🙌

**Reply by erquhart** [reply]
Yeah bookmarking that, really helpful breakdown!

I think I've always assumed lists without field arguments _could_ automatically update, I think that's where my primary confusion has been. Thanks again for the responses here.

**Reply by erquhart** [reply]
Almost seems like there's an opportunity here for an exchange that handles relationships, as my graphcache config has a lot of repetitive linking right now (and it's about to get a lot more) that I'm mostly handling with helper functions. Pointing it out here in case you already know why that's a bad idea. Otherwise I may look into that as a side project.

**Comment by nicolodavis** [comment]
I'm having the same problem.

After an `optimistic` is added, the regular update coming back from the API doesn't update the cache either.

**Reply by nicolodavis** [reply]
**Providing a bit more detail:**
1. I'm updating a query that's listing some entities using `cache.link` in the cache updater.
2. Everything works fine without optimistic updates.
3. The moment I add an optimistic update (which returns data that is identical to the mutation resolver) no update happens .

**After further investigation, this is what I observed:**
When the cache updater is invoked after the optimistic update, the new entity doesn't seem to exist in the cache at the point `cache.link` is invoked to update the list. So, in effect, it creates a dangling pointer that I suspect is the source of further problems.

Sure enough, using `cache.writeFragment` to manually create the new entity first before using `cache.link` to update the list **fixes the problem**.

Is there a reason for the difference in behavior in the updater when invoked after the optimistic update compared to after the mutation resolver (even though the data returned is identical in both cases)? I would have expected all the entities in the mutation result to have been added to the cache at the point the updater is called.

**Reply by nicolodavis** [reply]
For reference, I've designed my app in such a way that all mutations return data in the format:

```
{
  additions: [ ... ],
  updates: [ ... ],
  deletions: [ ... ],
}
```

I use this to create a generic updater that is able to work across all mutations and update the cache automatically without me needing to craft updaters manually. The cache does not seem to have a problem parsing entities within those individual lists after the resolver returns, so I'm not sure why things don't work automatically when invoked after an optimistic update.

**Comment by mtourj** [comment]


**Reply by JoviDeCroock** [reply]
I mean, your mutation is asking for an `id` field but your return of your optimistic isn't giving it back that is expected 😅 

```js
        optimistic: {
          createComment: (vars) => ({
            __typename: 'Comment',
            id: 'temp-optimistic-' + Math.random(),
            body: vars.body,
          }),
        },
```

This is also documented https://formidable.com/open-source/urql/docs/graphcache/cache-updates/#optimistic-updates

> It's important to ensure that our optimistic mutations return all data that the real mutation may return. If our mutations request a field in their selection sets that our optimistic mutation doesn't contain then we'll see a warning, since this is a common mistake. To work around not having enough data we may use methods like cache.readFragment and cache.resolve to retrieve more data from our cache.

**Reply by mtourj** [reply]
Oops. Sorry, am tired. Thanks lol
